File: app/src/main/java/com/stedi/randomimagegenerator/app/view/activity/HomeActivity.java
package com.stedi.randomimagegenerator.app.view.activity;

import android.Manifest;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.squareup.otto.Subscribe;
import com.stedi.randomimagegenerator.ImageParams;
import com.stedi.randomimagegenerator.app.R;
import com.stedi.randomimagegenerator.app.di.modules.HomeModule;
import com.stedi.randomimagegenerator.app.di.qualifiers.RootSavePath;
import com.stedi.randomimagegenerator.app.model.data.Preset;
import com.stedi.randomimagegenerator.app.other.CachedBus;
import com.stedi.randomimagegenerator.app.other.Utils;
import com.stedi.randomimagegenerator.app.other.logger.Logger;
import com.stedi.randomimagegenerator.app.presenter.interfaces.HomePresenter;
import com.stedi.randomimagegenerator.app.view.activity.base.BaseActivity;
import com.stedi.randomimagegenerator.app.view.adapters.PresetsAdapter;
import com.stedi.randomimagegenerator.app.view.components.GeneratorTypeImageLoader;
import com.stedi.randomimagegenerator.app.view.components.ListSpaceDecoration;
import com.stedi.randomimagegenerator.app.view.dialogs.ConfirmDialog;
import com.stedi.randomimagegenerator.app.view.dialogs.GenerationDialog;

import java.io.File;
import java.io.Serializable;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class HomeActivity extends BaseActivity implements HomePresenter.UIImpl, PresetsAdapter.ClickListener {
    private static final String KEY_HOME_PRESENTER_STATE = "KEY_HOME_PRESENTER_STATE";
    private static final int REQUEST_CODE_WRITE_EXTERNAL = 123;
    private static final int REQUEST_CONFIRM_DELETE = 321;
    private static final int REQUEST_CONFIRM_GENERATE = 231;

    @Inject HomePresenter presenter;
    @Inject Logger logger;
    @Inject CachedBus bus;
    @Inject GeneratorTypeImageLoader adapterImageLoader;
    @Inject @RootSavePath String rootSavePath;

    @BindView(R.id.home_activity_recycler_view) RecyclerView recyclerView;
    @BindView(R.id.home_activity_empty_view) View emptyView;
    @BindView(R.id.home_activity_fab) FloatingActionButton fab;

    private PresetsAdapter adapter;
    private Preset startGenerationPreset;
    private String confirmPresetName;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        getActivityComponent().plus(new HomeModule()).inject(this);
        super.onCreate(savedInstanceState);

        setContentView(R.layout.home_activity);
        ButterKnife.bind(this);

        if (savedInstanceState != null) {
            Serializable state = savedInstanceState.getSerializable(KEY_HOME_PRESENTER_STATE);
            if (state != null)
                presenter.onRestore(state);
        }

        fab.hide(fabShowHideListener);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setHasFixedSize(true);
        recyclerView.addItemDecoration(new ListSpaceDecoration(
                Utils.dp2pxi(this, R.dimen.common_v_spacing), Utils.dp2pxi(this, R.dimen.common_lr_spacing)));
        adapter = new PresetsAdapter(adapterImageLoader, rootSavePath, this);
        recyclerView.setAdapter(adapter);
        recyclerView.addOnScrollListener(recyclerScrollListener);
    }

    @Override
    protected void onStart() {
        super.onStart();
        presenter.onAttach(this);
        bus.register(this);
        presenter.fetchPresets();
    }

    @Override
    protected void onStop() {
        super.onStop();
        bus.unregister(this);
        presenter.onDetach();
    }

    @Override
    public void onPresetsFetched(@Nullable Preset pendingPreset, @NonNull List<Preset> presets) {
        fab.show(fabShowHideListener);
        adapter.set(presets, pendingPreset);
        refreshEmptyView();
    }

    @Override
    public void onFailedToFetchPresets() {
        fab.show(fabShowHideListener);
        Utils.toastLong(this, R.string.failed_fetch_presets);
        refreshEmptyView();
    }

    @OnClick(R.id.home_activity_fab)
    public void onFabClick(View v) {
        GenerationStepsActivity.startActivity(this, true);
    }

    @Override
    public void onCardClick(@NonNull Preset preset) {
        presenter.editPreset(preset);
    }

    @Override
    public void showEditPreset() {
        GenerationStepsActivity.startActivity(this, false);
    }

    @Override
    public void onSaveClick(@NonNull Preset preset) {
        presenter.editPreset(preset);
    }

    @Override
    public void onDeleteClick(@NonNull Preset preset) {
        confirmPresetName = preset.getName();
        presenter.deletePreset(preset);
    }

    @Override
    public void onPresetDeleted(@NonNull Preset preset) {
        logger.log(this, "onPresetDeleted " + preset);
        adapter.remove(preset);
        refreshEmptyView();
    }

    @Override
    public void onFailedToDeletePreset() {
        Utils.toastLong(this, R.string.failed_delete_preset);
    }

    @Override
    public void onGenerateClick(@NonNull Preset preset) {
        if (checkForPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, REQUEST_CODE_WRITE_EXTERNAL)) {
            confirmPresetName = preset.getName();
            presenter.startGeneration(preset);
        } else {
            startGenerationPreset = preset;
        }
    }

    @Override
    public void showConfirmLastAction(@NonNull HomePresenter.Confirm confirm) {
        if (confirm == HomePresenter.Confirm.DELETE_PRESET) {
            ConfirmDialog.newInstance(REQUEST_CONFIRM_DELETE,
                    getString(R.string.confirm_action), getString(R.string.are_you_sure_delete_s_preset, confirmPresetName))
                    .show(getSupportFragmentManager(), ConfirmDialog.class.getSimpleName());
        } else if (confirm == HomePresenter.Confirm.GENERATE_FROM_PRESET) {
            ConfirmDialog.newInstance(REQUEST_CONFIRM_GENERATE,
                    getString(R.string.confirm_action), getString(R.string.are_you_sure_generate_preset_s, confirmPresetName))
                    .show(getSupportFragmentManager(), ConfirmDialog.class.getSimpleName());
        }
    }

    @Subscribe
    public void onPermissionEvent(PermissionEvent event) {
        if (event.requestCode == REQUEST_CODE_WRITE_EXTERNAL) {
            if (event.isGranted && startGenerationPreset != null) {
                confirmPresetName = startGenerationPreset.getName();
                //noinspection MissingPermission
                presenter.startGeneration(startGenerationPreset);
            }
            startGenerationPreset = null;
        }
    }

    @Subscribe
    public void onConfirmDialogCallback(ConfirmDialog.Callback callback) {
        logger.log(this, "onConfirmDialogCallback " + callback.confirm);
        if (callback.requestCode == REQUEST_CONFIRM_DELETE ||
                callback.requestCode == REQUEST_CONFIRM_GENERATE) {
            if (callback.confirm) {
                presenter.confirmLastAction();
            } else {
                presenter.cancelLastAction();
            }
        }
    }

    @Override
    public void onStartGeneration() {
        logger.log(this, "onStartGeneration");
        GenerationDialog.getInstance(getSupportFragmentManager()).onStartGeneration();
    }

    @Override
    public void onGenerated(@NonNull ImageParams imageParams, @NonNull File imageFile) {
        logger.log(this, "onGenerated");
        sendBroadcast(new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(imageFile)));
        GenerationDialog.getInstance(getSupportFragmentManager()).onGenerated(imageParams, imageFile);
    }

    @Override
    public void onGenerationUnknownError() {
        logger.log(this, "onGenerationUnknownError");
        GenerationDialog.getInstance(getSupportFragmentManager()).onGenerationUnknownError();
    }

    @Override
    public void onFailedToGenerate(@NonNull ImageParams imageParams) {
        logger.log(this, "onFailedToGenerate");
        GenerationDialog.getInstance(getSupportFragmentManager()).onFailedToGenerate(imageParams);
    }

    @Override
    public void onFinishGeneration() {
        logger.log(this, "onFinishGeneration");
        GenerationDialog.getInstance(getSupportFragmentManager()).onFinishGeneration();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(KEY_HOME_PRESENTER_STATE, presenter.onRetain());
    }

    private void refreshEmptyView() {
        emptyView.setVisibility(adapter.getItemCount() == 0 ? View.VISIBLE : View.GONE);
    }

    private RecyclerView.OnScrollListener recyclerScrollListener = new RecyclerView.OnScrollListener() {
        @Override
        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            if (dy > 0) {
                fab.hide(fabShowHideListener);
            } else {
                fab.show(fabShowHideListener);
            }
        }
    };

    private FloatingActionButton.OnVisibilityChangedListener fabShowHideListener = new FloatingActionButton.OnVisibilityChangedListener() {
        @Override
        public void onShown(FloatingActionButton fab) {
            fab.setVisibility(View.VISIBLE);
        }

        @Override
        public void onHidden(FloatingActionButton fab) {
            fab.setVisibility(View.GONE);
        }
    };
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/view/activity/HomeActivity.kt
package com.stedi.randomimagegenerator.app.view.activity

import android.Manifest
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.squareup.otto.Subscribe
import com.stedi.randomimagegenerator.app.R
import com.stedi.randomimagegenerator.app.model.data.Preset
import com.stedi.randomimagegenerator.app.other.LockedBus
import com.stedi.randomimagegenerator.app.other.dim2px
import com.stedi.randomimagegenerator.app.other.showToast
import com.stedi.randomimagegenerator.app.presenter.interfaces.HomePresenter
import com.stedi.randomimagegenerator.app.view.activity.base.BaseActivity
import com.stedi.randomimagegenerator.app.view.adapters.PresetsAdapter
import com.stedi.randomimagegenerator.app.view.components.BaseViewModel
import com.stedi.randomimagegenerator.app.view.components.ListSpaceDecoration
import com.stedi.randomimagegenerator.app.view.dialogs.ConfirmDialog
import com.stedi.randomimagegenerator.app.view.dialogs.GenerationDialog
import javax.inject.Inject

class HomeActivityModel : BaseViewModel<HomeActivity>() {
    @Inject lateinit var presenter: HomePresenter
    @Inject lateinit var bus: LockedBus

    override fun onCreate(view: HomeActivity) {
        view.activityComponent.inject(this)
    }

    override fun onCleared() {
        super.onCleared()
        presenter.onDestroy()
    }
}

class HomeActivity : BaseActivity(), HomePresenter.UIImpl, PresetsAdapter.ClickListener {
    private val KEY_PRESENTER_STATE = "KEY_PRESENTER_STATE"
    private val KEY_PERMISSION_PRESET = "KEY_PERMISSION_PRESET"

    private val REQUEST_CODE_WRITE_EXTERNAL = 123
    private val REQUEST_CONFIRM_DELETE = 321
    private val REQUEST_CONFIRM_GENERATE = 231

    private lateinit var viewModel: HomeActivityModel

    @BindView(R.id.home_activity_recycler_view) lateinit var recyclerView: RecyclerView
    @BindView(R.id.home_activity_empty_view) lateinit var emptyView: View
    @BindView(R.id.home_activity_fab) lateinit var fab: FloatingActionButton

    private lateinit var adapter: PresetsAdapter

    private var permissionPreset: Preset? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel = ViewModelProviders.of(this).get(HomeActivityModel::class.java)
        val restorePresenter = !viewModel.isInitialized && savedInstanceState != null
        viewModel.init(this)

        setContentView(R.layout.home_activity)
        ButterKnife.bind(this)

        savedInstanceState?.apply {
            if (restorePresenter) {
                viewModel.presenter.onRestore(getSerializable(KEY_PRESENTER_STATE))
            }
            permissionPreset = getParcelable(KEY_PERMISSION_PRESET)
        }

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.setHasFixedSize(true)
        recyclerView.addItemDecoration(ListSpaceDecoration(dim2px(R.dimen.common_v_spacing), dim2px(R.dimen.common_lr_spacing)))
        adapter = PresetsAdapter(this, this)
        recyclerView.adapter = adapter
        recyclerView.addOnScrollListener(recyclerScrollListener)
    }

    override fun onStart() {
        super.onStart()
        viewModel.presenter.onAttach(this)
        viewModel.bus.register(this)
        viewModel.presenter.fetchPresets()
    }

    override fun onStop() {
        super.onStop()
        viewModel.bus.unregister(this)
        viewModel.presenter.onDetach()
    }

    @OnClick(R.id.home_activity_fab)
    fun onFabClick(v: View) {
        viewModel.presenter.newPreset()
    }

    override fun onCardClick(preset: Preset) {
        viewModel.presenter.editPreset(preset)
    }

    override fun onDeleteClick(preset: Preset) {
        viewModel.presenter.deletePreset(preset)
    }

    override fun onGenerateClick(preset: Preset) {
        if (checkForPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, REQUEST_CODE_WRITE_EXTERNAL)) {
            viewModel.presenter.startGeneration(preset)
        } else {
            permissionPreset = preset
        }
    }

    override fun onSaveClick(preset: Preset) {
        viewModel.presenter.editPreset(preset)
    }

    override fun onPresetsFetched(pendingPreset: Preset?, presets: List<Preset>) {
        adapter.set(presets, pendingPreset)
        refreshEmptyView()
    }

    override fun onFailedToFetchPresets() {
        showToast(R.string.failed_fetch_presets)
        refreshEmptyView()
    }

    override fun showConfirmDeletePreset(preset: Preset) {
        ConfirmDialog.newInstance(REQUEST_CONFIRM_DELETE,
                getString(R.string.confirm_action), getString(R.string.are_you_sure_delete_s_preset, preset.name))
                .show(supportFragmentManager, ConfirmDialog::class.java.simpleName)
    }

    override fun onPresetDeleted(preset: Preset) {
        adapter.remove(preset)
        refreshEmptyView()
    }

    override fun onFailedToDeletePreset() {
        showToast(R.string.failed_delete_preset)
    }

    override fun showConfirmGeneratePreset(preset: Preset) {
        ConfirmDialog.newInstance(REQUEST_CONFIRM_GENERATE,
                getString(R.string.confirm_action), getString(R.string.are_you_sure_generate_preset_s, preset.name))
                .show(supportFragmentManager, ConfirmDialog::class.java.simpleName)
    }

    override fun showGenerationDialog(preset: Preset) {
        GenerationDialog.newInstance(preset).show(supportFragmentManager)
    }

    override fun showEditPreset() {
        GenerationStepsActivity.start(this, false)
    }

    override fun showCreatePreset() {
        GenerationStepsActivity.start(this, true)
    }

    @Subscribe
    fun onPermissionEvent(event: BaseActivity.PermissionEvent) {
        if (event.requestCode == REQUEST_CODE_WRITE_EXTERNAL) {
            permissionPreset?.apply {
                permissionPreset = null
                if (event.isGranted) {
                    viewModel.presenter.startGeneration(this)
                }
            }
        }
    }

    @Subscribe
    fun onConfirmDialogCallback(callback: ConfirmDialog.Callback) {
        if (callback.requestCode == REQUEST_CONFIRM_DELETE) {
            viewModel.presenter.confirmDeletePreset(callback.confirm)
        } else if (callback.requestCode == REQUEST_CONFIRM_GENERATE) {
            viewModel.presenter.confirmStartGeneration(callback.confirm)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable(KEY_PRESENTER_STATE, viewModel.presenter.onRetain())
        outState.putParcelable(KEY_PERMISSION_PRESET, permissionPreset)
    }

    private fun refreshEmptyView() {
        emptyView.visibility = if (adapter.itemCount == 0) View.VISIBLE else View.GONE
    }

    private val recyclerScrollListener = object : RecyclerView.OnScrollListener() {
        override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
            if (dy > 0) {
                fab.hide(fabShowHideListener)
            } else {
                fab.show(fabShowHideListener)
            }
        }
    }

    private val fabShowHideListener = object : FloatingActionButton.OnVisibilityChangedListener() {
        override fun onShown(fab: FloatingActionButton) {
            fab.visibility = View.VISIBLE
        }

        override fun onHidden(fab: FloatingActionButton) {
            fab.visibility = View.GONE
        }
    }
}